Business logic

N1ghtmarex opened the pull request.

N1ghtmarex committed:
Исправлена конфигурация Keycloak

N1ghtmarex committed:
Изменена роль по умолчанию при создании аккаунта пользователя

N1ghtmarex committed:
Добавлено получение профилей пользователей

N1ghtmarex committed:
Добавлено создание новой локации

N1ghtmarex committed:
Перемещены User View Models

N1ghtmarex committed:
Добавлено получение сведений о локациях

N1ghtmarex committed:
Добавлено получение списка адресов

N1ghtmarex committed:
Добавлено обновление локации

N1ghtmarex committed:
Добавлен HttpContextAccessor

N1ghtmarex committed:
Добавлен внешний идентификатор пользователя

N1ghtmarex committed:
Добавлена проверка на роль администратора

N1ghtmarex committed:
Реализовано добавление\удаление избранных локаций

N1ghtmarex committed:
Добавлен просмотр избранных локаций

N1ghtmarex committed:
Изменен endpoint

N1ghtmarex committed:
Добавлено получение пользователя в UserService

N1ghtmarex committed:
Добавлена публикация изображений локаций

N1ghtmarex committed:
Добавлено установление аватара локации

N1ghtmarex committed:
Добавлено отображение изображений локации

N1ghtmarex committed:
Исправлено сохранение автара локации

N1ghtmarex committed:
merge branch "UserAuthRegister" into "BusinessLogic"

N1ghtmarex committed:
Изменения после слияния

N1ghtmarex committed:
Добавлен комментарий для метода регистрация пользователя

v4dikos1 commented on Backend/Api/Controllers/LocationController.cs:
```diff
@@ -0,0 +1,102 @@
+        /// <param name="cancellationToken">Токен отмены</param>
+        /// <returns></returns>
+        [HttpPost("favorite/{LocationId}")]
+        public async Task<CreatedOrUpdatedEntityViewModel<Ulid>> AddOrRemoveFavoriteLocation([FromQuery] AddOrRemoveFavoriteLocationCommand command,
```
Как-будто бы все намешано в одном месте. Локации добавляет админ, но админ по идее не должен добавлять что-либо в избранное. И наоборот - обычный юзер не должен иметь возможности добавлять локации. Здесь надо во-первых добавить разделение по ролям, во-вторых - разбить контроллер на админскую часть и на юзерскую часть

N1ghtmarex replied:
Добавил AdminLocationController и вынес туда админские методы

N1ghtmarex replied:
> админ по идее не должен добавлять что-либо в избранное

Кстати, с этим не соглашусь, поскольку в моей логике админ - это тот же юзер, но с расширенными правами
Ну это уже придирки с моей стороны, наверное)

v4dikos1 commented on Backend/Api/Controllers/PictureController.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <param name="command">Модель запроса</param>
+        /// <param name="cancellationToken">Токен отмены</param>
+        /// <returns></returns>
+        [HttpPost("locationId/{LocationId}")]
```
Путь довольно странный

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Addresses/Handlers/AddressQueriesHandlers.cs:
```diff
@@ -0,0 +1,42 @@
+        public async Task<PagedResult<AddressListViewModel>> Handle(GetAddressesListQuery request, CancellationToken cancellationToken)
+        {
+            var addressQuery = dbContext.Addresses
+                .OrderBy(x => x.Region)
```
Рекомендую добавлять .AsNoTracking() в queriesHandler'ы
Это даст:
1) Неизменяемость данных (даже если что-то поменяешь и сохранишь изменения, то ничего не сохранится), а именно ее мы хотим для query в CQRS
2) Ускорение запроса, т.к EF больше не будет трекать сущности

N1ghtmarex replied:
исправил

v4dikos1 commented on Backend/Application/Addresses/Handlers/AddressQueriesHandlers.cs:
```diff
@@ -0,0 +1,42 @@
+
+            var result = await addressQuery
+                .ApplyPagination(request)
+                .Select(x => new AddressListViewModel
```
Мусорный код, который можно засунуть в маппер (Project на IQueryable)

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationCommandsHandler.cs:
```diff
@@ -0,0 +1,193 @@
+                throw new ForbiddenException("Только администратор может добавлять новые локации!");
+            }
+
+            var address = await dbContext.Addresses
```
Страшно, очень страшно. Разве нет каких-нибудь уникальных идентификаторов для адресов, по которым бы можно было определять адрес?

v4dikos1 replied:
И если уж делаешь SingleOrDefault, то нужно на сущность навесить constraint по всем этим полям на уникальность

N1ghtmarex replied:
Удалил адрес, теперь используются координаты

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationCommandsHandler.cs:
```diff
@@ -0,0 +1,193 @@
+
+            if (address == null)
+            {
+                var addressToCreate = new Address
```
вынести в маппер, сильно засоряет код бизнес логики

N1ghtmarex replied:
Исправил

N1ghtmarex replied:
Удалил адрес, теперь используются координаты

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationCommandsHandler.cs:
```diff
@@ -0,0 +1,193 @@
+            }
+
+            var existLocation = await dbContext.Locations
+                .SingleOrDefaultAsync(x => x.Name == request.Body.Name || x.Address == address, cancellationToken);
```
Названия case-sensitive?

N1ghtmarex replied:
добавил приведение к нижнему регистру

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationCommandsHandler.cs:
```diff
@@ -0,0 +1,193 @@
+                throw new ObjectExistsException("Локация с таким названием или адресом уже существует!");
+            }
+
+            var locationToCreate = new Location
```
маппер

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationCommandsHandler.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            var user = await userService.GetUserByExternalIdAsync(httpContext.IdentityUserId, cancellationToken);
+
+            if (user.Role != UserRole.Admin)
```
Если вынести валидацию роли на уровень контроллера ([Authorize(Roles="Admin")]), то не придется каждый раз делать проверки в бизнес логике

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationCommandsHandler.cs:
```diff
@@ -0,0 +1,193 @@
+                address = createdAddress.Entity;
+            }
+
+            existsLocation.Name = request.Body.Name;
```
В маппер

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationQueriesHandler.cs:
```diff
@@ -0,0 +1,131 @@
+        public async Task<LocationViewModel> Handle(GetLocationQuery request, CancellationToken cancellationToken)
+        {
+            var location = await dbContext.Locations
+                .Include(x => x.Address)
```
.AsNoTracking

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationQueriesHandler.cs:
```diff
@@ -0,0 +1,131 @@
+        {
+            var location = await dbContext.Locations
+                .Include(x => x.Address)
+                .Select(x => new LocationViewModel
```
Маппинг (проекция)

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationQueriesHandler.cs:
```diff
@@ -0,0 +1,131 @@
+
+            var result = await locationQuery
+                .ApplyPagination(request)
+                .Select(x => new LocationListViewModel
```
маппинг

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationQueriesHandler.cs:
```diff
@@ -0,0 +1,131 @@
+
+        public async Task<PagedResult<LocationListViewModel>> Handle(GetLocationsListQuery request, CancellationToken cancellationToken)
+        {
+            var locationQuery = dbContext.Locations
```
AsNoTracking

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Locations/Handlers/LocationQueriesHandler.cs:
```diff
@@ -0,0 +1,131 @@
+            return result.AsPagedResult(request, await locationQuery.CountAsync(cancellationToken));
+        }
+
+        public async Task<PagedResult<LocationListViewModel>> Handle(GetFavoriteLocationListQuery request, CancellationToken cancellationToken)
```
Аналогично

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Application/Users/Handlers/UserQueriesHandler.cs:
```diff
@@ -0,0 +1,51 @@
+    {
+        public async Task<UserViewModel> Handle(GetUserQuery request, CancellationToken cancellationToken)
+        {
+            var user = await dbContext.Users
```
AsNoTracking
мапперы

N1ghtmarex replied:
Исправил

v4dikos1 commented on Backend/Domain/EntityConfigurations/AddressConfiguration.cs:
```diff
@@ -9,7 +9,12 @@ internal class AddressConfiguration : IEntityTypeConfiguration<Address>
         public void Configure(EntityTypeBuilder<Address> builder)
         {
             builder.ToTable("address");
```
Маловато конфигурации, с учетом того, как ты производишь поиск и добавление адресов. 
Как минимум нужны констрейнты на уникальность, индексы для быстрого поиска

v4dikos1 replied:
Явное лучше неявного. Лучше явно указать, какие поля обязательные, какие - нет. Указать максимально допустимую длину и т.д.

v4dikos1 replied:
Если не добавлять констрейнты на уровне бд, твой код очень легко сломать. Да, у тебя есть проверки на уникальность имени или других полей в бизнес логике при создании сущности. Но это не дает полных гарантий, что в бд не запишется две одинаковых записи. А вот ограничение на уровне бд как раз дает

N1ghtmarex replied:
Удалил адрес, теперь используются координаты

v4dikos1 commented on Backend/Domain/EntityConfigurations/LocationConfiguration.cs:
```diff
@@ -21,9 +21,15 @@ public void Configure(EntityTypeBuilder<Location> builder)
 
```
Аналогично. Не хватает конфигурации - индексов, констрейнтов

N1ghtmarex replied:
изменил конфигурацию

v4dikos1 commented on Backend/Infrastructure/Services/FileService.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Infrastructure.Services
+{
+    public class FileService : IFileService
```
Может все-таки S3?
Максим вон тебе minio развернет и с кайфом. Хранить файлы в файловом хранилище как-то уже не очень современно что-ли

v4dikos1 replied:
Ну и раз уж заговорили про S3, то можно избавиться от лишней нагрузки на сервер - можно грузить файлы прямиком с фронта в s3, минуя лишнюю прослойку в виде сервера. На стороне сервера достаточно реализовать выдачу подписанных ссылок на загрузку/получение файлов

N1ghtmarex replied:
Настроил хранение файлов в MinIO

N1ghtmarex committed:
refactor: - Изменен url метода публикации изображения; для LocationId добавлен атрибут [FromQuery]

N1ghtmarex committed:
refactor: оптимизация запросов
* добавлено AsNoTracking() в Handler'ы запросов

N1ghtmarex committed:
feat: Добавлены мапперы

N1ghtmarex committed:
fix: Исправлен маппинг адреса

N1ghtmarex committed:
feat: Добавлен общий маппер
+ Добавлен общий маппер
* NamedMapping вынесены в общий маппер

N1ghtmarex committed:
fix: Исправлен вызов методов генерации полей

N1ghtmarex committed:
feat: Добавлен маппер
+ Добавлен маппер для сущности User

N1ghtmarex committed:
feat: В маппер добавлен новый метод
+ Добавлен метод ProjectToListViewModel()

N1ghtmarex committed:
refactor: Добавлено использование маппера
* При выборке адресов добавлен вызов ProjectToListViewModel()

N1ghtmarex committed:
refactor: Изменен маппер
+ В AddressMapper добавлен метод маппинга в сущность
+ В GeneralMapper добавлен метод генерации Ulid в виде строки

N1ghtmarex committed:
refactor!: Добавлено использование маппера
* Добавлено использование метода .MapToEntity()

N1ghtmarex committed:
Revert "refactor!: Добавлено использование маппера"

This reverts commit 9e26a63835aabe5253d7851975cf5be3b5a77514.

N1ghtmarex committed:
Reapply "refactor!: Добавлено использование маппера"

This reverts commit e888d5516ae581e575b3f6b60dc4c5da37119153.

N1ghtmarex committed:
fix: Исправлены мапперы
+ В AddressMapper и UserMapper добавлено использование методов генерации Id

N1ghtmarex committed:
refactor!: Добавлено использование мапперов

N1ghtmarex committed:
fix: Исправление маппера пользователей
+ Добавлен маппинг поля CreatedAt

N1ghtmarex committed:
fix: Исправление маппинга
+ Добавлен маппинг изображений
+ Добавлен маппинг избранных локаций
* В запросах добавлен Include() изображений

N1ghtmarex committed:
refactor: Добавлено использование мапперов
* В запросы добавлено использование маппинга
+ Добавлен метод маппинга Picture в Entity и Bind
+ Добавлен метод маппинга UserFavorite в Entity

N1ghtmarex committed:
refactor: Использование маппинга
+ Добавлен метод маппинга при обновлении сущности Location

N1ghtmarex committed:
refactor: Изменено получение существующей локации при создании новой
* Добавлено приведение названий к нижнему регистру

N1ghtmarex committed:
refactor: разделение LocationController на админский и пользовательский
+ Добавлен AdminLocationController
* Из LocationController перенесены методы, требующие роль Admin в AdminLocationController

N1ghtmarex committed:
refactor: выполнена переработка базы данных
+ В сущность Location вместо адреса добавлены поля Latitude и Longitude
+ Выполнена миграция
- Удалена сущность адреса
- Удален AddressController
- Удален AddressMapper
- Удалена бизнес-логика, связанная с адресом
* В запросах удалено использование адреса

N1ghtmarex committed:
refactor: изменена конфигурация сущности Location
+ Добавлены индексы для полей Id, Type, CreatedAt, UpdatedAt, IsArchive
+ Добавлено явное указание уникальности для поля Id
+ Добавлено ограничение максимальной длины названия локации в 50 символов
+ Добавлено ограничение максимальной длины описания локации в 100 символов
+ Выполнена миграция

N1ghtmarex committed:
refactor: Изменена проверка роли
+ В PictureController и LocationController добавлен атрибут Roles = "Admin"
- В хендлерах убрана проверка роли
* Изменен способ получения ролей из токена

N1ghtmarex committed:
feat: Добавлена интеграция c MinIO
+ Добавлена конфигурация MinIO
* Изменен FileService: теперь файлы сохраняются в MinIO
